### PR TITLE
Add generic secret handling for deploy and update

### PR DESCRIPTION
**Pull request by LucasRoesler**
## Description
- During deploy, inspect the secret type and mount docker config secrets
as image pull secrets and all other secrets as files
- fix import name in the tests

## Motivation and Context
This address issue #82 

- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

Tested with the following script to mount a function with multiple secrets and validate that they are mounted in the correct location with the expected content

```sh
#!/bin/bash
#
# Usage: test.sh file1.txt file2.txt

kubectl create secret generic secret1 --from-file=$1
kubectl create secret generic secret2 --from-file=$2

echo "Deploy test function"
curl \
    -X POST \
    -d '{"image": "functions/alpine:latest", "envProcess": "ls -al /run/secrets", "service": "ls-secret-func", "secrets": ["secret1", "secret2"]}' \
    $(minikube ip):31111/system/functions

curl \
    -X POST \
    -d '{"image": "functions/alpine:latest", "envProcess": "cat /run/secrets/secretone.txt", "service": "cat-secretone", "secrets": ["secret1", "secret2"]}' \
    $(minikube ip):31111/system/functions

curl \
    -X POST \
    -d '{"image": "functions/alpine:latest", "envProcess": "cat /run/secrets/secrettwo.txt", "service": "cat-secrettwo", "secrets": ["secret1", "secret2"]}' \
    $(minikube ip):31111/system/functions

echo "Wait a moment"
sleep 5s

echo "Call the function"
curl \
    -X POST \
    $(minikube ip):31111/function/ls-secret-func
curl \
    -X POST \
    $(minikube ip):31111/function/cat-secretone
curl \
    -X POST \
    $(minikube ip):31111/function/cat-secrettwo

echo "Remove the function"
curl \
    -X DELETE \
    -d '{"functionName": "ls-secret-func"}' \
    $(minikube ip):31111/system/functions
curl \
    -X DELETE \
    -d '{"functionName": "cat-secretone"}' \
    $(minikube ip):31111/system/functions
curl \
    -X DELETE \
    -d '{"functionName": "cat-secrettwo"}' \
    $(minikube ip):31111/system/functions

kubectl delete secret secret1
kubectl delete secret secret2
```

Create two files `file1.txt` and `file2.txt` with different content.  Then, save the above script at `test.sh` and run `./test.sh file1.txt file2.txt`. it will create the k8s secrets from the provided files, create test functions that reference those secrets, call those functions, and then clean up. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
This looks sensible, slightly more work than expected for creating "generic secrets" due to the volume mounts. We should ask for some testers in the community now.

**Comment by LucasRoesler** [comment]
I found an issue with this implementation when a user specifies multiple secrets

```
spec.template.spec.containers[0].volumeMounts[1].mountPath: Invalid value: "/run/secrets": must be unique
```

The problem is caused by adding 

```go
VolumeMount{
	Name:      secretName,
	ReadOnly:  true,
	MountPath: "/run/secrets",
}
```

for each generic secret.  I was originally expecting k8s to merge the mounts into the same folder, but  that apparently does not work.  Perhaps someone with more k8s experience can offer a suggestion. But right now, I don't see an obvious way to mount secrets into the same folder and then mimic the behavior of Swarm.

**Comment by LucasRoesler** [comment]
I just updated the change to use a projected volume so that all of the specified secrets are mounted into `/run/secrets`, I have also included testing instructions.

**Comment by alexellis** [comment]
Glad you found a way to do this Lucas. I'll be able to look into this mid next week.

**Comment by ericstoekl** [comment]
Testing this out and it looks great. Ran the sample test script (in the original post) and the `secretone` and `secrettwo` functions work as expected.

The following is related to `faas-cli` integration, not this PR per se

@alexellis I was trying to deploy a function with secrets using `faas-cli`, by including the `secrets:` function in the stack.yml, but it won't work at this time. This is because the `stack/schema.go` file does not include a `Secrets` data member, so the stack.yml may contain secrets but they will not get parsed out.

https://github.com/openfaas/faas-cli/blob/5a05f302d309a530040f7ef7dc7aebe10b0f2bb9/stack/schema.go#L14

There is a [commit that added secrets here](https://github.com/openfaas/faas-cli/commit/b917264241c4ad80cbc74edd684f06ca3bed1348), but did not include this change which I think is necessary.

**Comment by alexellis** [comment]
So did you manage to configure `minikube` for RBAC successfully?

**Comment by LucasRoesler** [comment]
I just now got a firm validation with a vagrant provisioned kubernetes and RBAC. 

I used following project to define the kubernetes cluster with two nodes: https://github.com/xbernpa/vagrant-kubernetes-lab  My only adjustment to the Vagrantfile was adding my code directly as a synced folder.  Once that directory was accesible.  I saved the dev builds of `gateway` and `faas-netes` and loaded them on the `k8smaster` and `k8sworker` nodes.

With those images ready, I also needed to update my kube config `~/.kube/config` to include the users and clustes in the resulting `kubeconfig/admin.conf` that the Vagrant up command created. 

I then update the `faas.yml` file to reference those images and use the `IfNotPresent` pull policy.  With that inplace, I deployed openfaas using `kubectl apply -f ./faas.yml,monitoring.yml,rbac.yml`

With this in place, I was able to update the test script to point at the vagrant machine and everything worked as expected.  I am not 100% how to definitely ensure that the RBAC is being enforced correctly, but we can see that the expected permissions are given to the faas-controller

```sh
➜  faas-netes git:(feature-improved-secrets-handling) ✗ kubectl describe clusterrole faas-controller
Name:		faas-controller
Labels:		<none>
Annotations:	kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"rbac.authorization.k8s.io/v1beta1","kind":"ClusterRole","metadata":{"annotations":{},"name":"faas-controller","namespace":""},"rules":[{...
PolicyRule:
  Resources			Non-Resource URLs	Resource Names	Verbs
  ---------			-----------------	--------------	-----
  deployments.extensions	[]			[]		[get list watch create delete update]
  secrets			[]			[]		[get list watch]
  services			[]			[]		[get list watch create delete update]

➜  faas-netes git:(feature-improved-secrets-handling) ✗ kubectl describe clusterrolebinding faas-controller
Name:		faas-controller
Labels:		<none>
Annotations:	kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"rbac.authorization.k8s.io/v1beta1","kind":"ClusterRoleBinding","metadata":{"annotations":{},"name":"faas-controller","namespace":""},"ro...
Role:
  Kind:	ClusterRole
  Name:	faas-controller
Subjects:
  Kind			Name		Namespace
  ----			----		---------
  ServiceAccount	faas-controller	default
```

**Comment by ericstoekl** [comment]
I just tested these changes by using the same Vagrant repo mentioned above (https://github.com/xbernpa/vagrant-kubernetes-lab). I took a simpler approach to testing this out. Here's what I did:

1. `vagrant up`, master and worker nodes are created in Virtualbox VMs
2. `vagrant ssh k8smaster`, inside the master VM I deploy the faas-netes image built from this PR and pushed to my docker hub
3. Create a test container with image pull secrets, and a container with app secrets, following the guide here: https://github.com/LucasRoesler/faas-netes/blob/caeec32c2320b9f3b3d55ca098e8a57d1d33f443/SECRETS.md

Both the Image Pull Secrets and Application Secrets work. Confirmed the Image Pull Secrets as working because the private docker image deployed only when the image pull secret is set. Confirmed the App Secrets is working by checking that the files embedded in the secret I specified at deployment shows up under `/run/secrets/<filename>`.

I also tried using this without the commit that adds settings to `rbac.yml`, and it does not work in that case, so I think the `rbac.yml` changes are good as well.

Overall, 👍

**Comment by LucasRoesler** [comment]
@alexellis I just pushed a fresh rebase to resolve any merge conflicts

**Comment by alexellis** [comment]
@LucasRoesler @ericstoekl @stefanprodan I'm changing the RBAC to least privilege. Please could you try testing your PR along with the "yaml" branch in #105 ?

Excited to get this change merged along with the Docker Swarm PR you worked on too. 💯

File: SECRETS.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 ## Managing Secrets
 
+Secrets come in two flavors, image-pull secrets and application secrets.  Kubernetes uses [image-pull secrets to facilitate pulling docker images from a private registry](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).  Application secrets are simply sensitive values that your function requires to function, e.g. a database password. `faas-netes` supports both types of secrets.  If you are familiar with the inner workings of Kubernetes, then image-pull secrets will be used as you would expect from the Kubernetes documentation. Generic application secrets will behave in the same way that secrets are handled in Docker Swarm, specifically, they will be mounted as files in the `/run/secrets` folder.  This allows your functions to be agnostic of the deployment infrastructure. The relevant Kubernetes documentation for how this is implemented can be found [here](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod).
```
Comment (alexellis) [inline SECRETS.md]: Capitalise docker -> Docker

Prefer plain English:

> Kubernetes uses [image-pull secrets to facilitate pulling docker images from a private registry

> Kubernetes uses [image-pull secrets to pull images from private registries]

Can this be broken up onto multiple lines?

Comment (LucasRoesler) [reply SECRETS.md]: Just pushed a change that should clean this up.

**Comment by alexellis** [comment]
We should sync these changes into my new PR to re-organise the YAML over at https://github.com/openfaas/faas-netes/pull/105

**Comment by alexellis** [comment]
Merged! Thanks so much for seeing this through @LucasRoesler - it's making a very big impact to the project. ⭐️ 💯 🎉

**Pull request merged**